tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_packing.cc

```diff
@@ -0,0 +1,476 @@
+class BufferPacking : BufferPlacementTransformationBase {
+
+public:
+  template <typename StrategyT>
+  BufferPacking(Operation *op, StrategyT strategy)
+      : BufferPlacementTransformationBase(op),
+        userangeAnalysis(op, allocs, aliases), dominators(op) {
+    DenseMap<size_t, size_t> bufferGen;
+    std::vector<PackedInfo> packingInfos;
+    strategy.optimze(allocs, userangeAnalysis, packingInfos, bufferGen);
+
+    DenseMap<size_t, Value> generatedBuffers;
+
+    // Find common dominators.
+    Block *block = findAllocationsDominator(packingInfos);
+    // Find alloc position operation.
+    mlir::OpBuilder packBuilder(&(block->front()));
+    auto location = block->front().getLoc();
+
+    // Create the packed AllocOps.
+    for (auto &entry : bufferGen) {
+      auto memrefType =
+          MemRefType::get({entry.second}, packBuilder.getIntegerType(8));
+      Value packedAlloc =
+          packBuilder.create<memref::AllocOp>(location, memrefType);
+      generatedBuffers.insert(std::make_pair(entry.first, packedAlloc));
+    }
+
+    for (auto &packInfo : packingInfos) {
+      Value currentAlloc = packInfo.source;
+      Value targetBuffer = generatedBuffers[packInfo.bufferId];
+      size_t offset = packInfo.offset;
+      Operation *viewDefOp = currentAlloc.getDefiningOp();
+      Location loc = viewDefOp->getLoc();
+      mlir::OpBuilder viewBuilder(viewDefOp);
+
+      // Create a ConstantOp with the aligned offset.
+      Value constantOp = viewBuilder.create<mlir::ConstantOp>(
+          loc, viewBuilder.getIndexType(),
+          viewBuilder.getIntegerAttr(viewBuilder.getIndexType(), offset));
+
+      // Store the operands for the ViewOp.
+      SmallVector<Value, 4> newOperands{targetBuffer};
+      newOperands.push_back(constantOp);
+      ShapedType shape = currentAlloc.getType().cast<ShapedType>();
+
+      // Create a ViewOp with the shape of the old alloc and use the created
+      // packed alloc and the constant for the operands.
+      Value viewOp = viewBuilder.create<memref::ViewOp>(
+          loc, shape.cast<MemRefType>(), newOperands);
+
+      // Replace all old allocs references with the created ViewOp and
+      // afterwards remove the old allocs.
+      currentAlloc.replaceAllUsesWith(viewOp);
+      viewDefOp->erase();
+    }
+  }
+
+private:
+  UserangeAnalysis userangeAnalysis;
+  /// The current dominance info.
+  DominanceInfo dominators;
+
+  /// Find the block that dominates all buffer allocations.
+  Block *findAllocationsDominator(const std::vector<PackedInfo> &packingInfos) {
+    SmallPtrSet<Value, 16> allocValues;
+    for (auto &packInfo : packingInfos) {
+      allocValues.insert(packInfo.source);
+    }
+
+    // Find common dominators.
+    return findCommonDominator(packingInfos.begin()->source, allocValues,
+                               dominators);
+  }
+};
+
+/// Reuses already allocated buffer to save allocation operations.
```

Comment:
Comment is off. To land this, please move to a utility file.

Reply:
We move the function to a utils file.